[Merged by Bors] - Implement SSZ union type

paulhauner opened the pull request:
## Issue Addressed

NA

## Proposed Changes

Implements the "union" type from the SSZ spec for `ssz`, `ssz_derive`, `tree_hash` and `tree_hash_derive` so it may be derived for `enums`:

https://github.com/ethereum/consensus-specs/blob/v1.1.0-beta.3/ssz/simple-serialize.md#union

The union type is required for the merge, since the `Transaction` type is defined as a single-variant union `Union[OpaqueTransaction]`.

### Crate Updates

This PR will (hopefully) cause CI to publish new versions for the following crates:

- `eth2_ssz_derive`: `0.2.1` -> `0.3.0`
- `eth2_ssz`: `0.3.0` -> `0.4.0`
- `eth2_ssz_types`: `0.2.0` -> `0.2.1`
- `tree_hash`: `0.3.0` -> `0.4.0`
- `tree_hash_derive`: `0.3.0` -> `0.4.0`

These these crates depend on each other, I've had to add a workspace-level `[patch]` for these crates. A follow-up PR will need to remove this patch, ones the new versions are published.

### Union Behaviors

We already had SSZ `Encode` and `TreeHash` derive for enums, however it just did a "transparent" pass-through of the inner value. Since the "union" decoding from the spec is in conflict with the transparent method, I've required that all `enum` have exactly one of the following enum-level attributes:

#### SSZ

-  `#[ssz(enum_behaviour = "union")]`
    - matches the spec used for the merge
-  `#[ssz(enum_behaviour = "transparent")]`
    - maintains existing functionality
    - not supported for `Decode` (never was)
    
#### TreeHash

-  `#[tree_hash(enum_behaviour = "union")]`
    - matches the spec used for the merge
-  `#[tree_hash(enum_behaviour = "transparent")]`
    - maintains existing functionality

This means that we can maintain the existing transparent behaviour, but all existing users will get a compile-time error until they explicitly opt-in to being transparent.

### Legacy Option Encoding

Before this PR, we already had a union-esque encoding for `Option<T>`. However, this was with the *old* SSZ spec where the union selector was 4 bytes. During merge specification, the spec was changed to use 1 byte for the selector.

Whilst the 4-byte `Option` encoding was never used in the spec, we used it in our database. Writing a migrate script for all occurrences of `Option` in the database would be painful, especially since it's used in the `CommitteeCache`. To avoid the migrate script, I added a serde-esque `#[ssz(with = "module")]` field-level attribute to `ssz_derive` so that we can opt into the 4-byte encoding on a field-by-field basis.

The `ssz::legacy::four_byte_impl!` macro allows a one-liner to define the module required for the `#[ssz(with = "module")]` for some `Option<T> where T: Encode + Decode`.

Notably, **I have removed `Encode` and `Decode` impls for `Option`**. I've done this to force a break on downstream users. Like I mentioned, `Option` isn't used in the spec so I don't think it'll be *that* annoying. I think it's nicer than quietly having two different union implementations or quietly breaking the existing `Option` impl.

### Crate Publish Ordering

I've modified the order in which CI publishes crates to ensure that we don't publish a crate without ensuring we already published a crate that it depends upon.

## TODO

- [ ] Queue a follow-up `[patch]`-removing PR.

paulhauner committed:
Add basic encode impl

paulhauner committed:
Add UnionSelector to ssz encode

paulhauner committed:
Comment-out existing Option impl

paulhauner committed:
Add non-functional decode impl

paulhauner committed:
Complete Decode impl

paulhauner committed:
Tidy comments

paulhauner committed:
Remove option code

paulhauner committed:
Add comments to union selector

paulhauner committed:
Bump ssz and ssz_derive versions

paulhauner committed:
Remove tests

paulhauner committed:
Freshen cargo.lock

paulhauner committed:
Freshen cargo.lock

paulhauner committed:
Remove unused union impl for ssz encoder

paulhauner committed:
Add ssz tests

paulhauner committed:
Add top-level patch

paulhauner committed:
Add basic union tree hash method

paulhauner committed:
Tidy tree hash

paulhauner committed:
Fix bugs, bump version

paulhauner committed:
Add tree_hash_derive tests

paulhauner committed:
Patch ssz types

paulhauner committed:
Update tree hash to 0.4.0

paulhauner committed:
Finish tree hash tests

paulhauner committed:
Apply enum_behaviour

paulhauner committed:
Bump eth2_ssz_types

paulhauner committed:
Bump ssz, ssz_derive versions

paulhauner committed:
Add FourByteSszOption

paulhauner committed:
Apply ssz option to committee cache

paulhauner committed:
Add add_parameterized to Encoder

paulhauner committed:
Add "with" to Encode derive

paulhauner committed:
Add "with" for Decode

paulhauner committed:
Remove unused derive functions

paulhauner committed:
Fix bugs in ssz_derive

paulhauner committed:
Use only one line for ssz field attributes

paulhauner committed:
Add legacy mod to ssz

paulhauner committed:
Use legacy option in proto array

paulhauner committed:
Import super::* in legacy macro

paulhauner committed:
Add enum_behaviour = transparent to misc structs

paulhauner committed:
Remove FourByteSszOption

paulhauner committed:
Resolve old uses of Option for SSZ

paulhauner committed:
Move derive tests

paulhauner committed:
Re-order crate publish

paulhauner committed:
Remove ssz_types dep from tree_hash

paulhauner committed:
Publish ssz before tree hash

paulhauner committed:
Fix Vec<u8> variant decode derive issue

paulhauner committed:
Tidy comments

paulhauner committed:
Reinstate legacy option tests, tidy macro

paulhauner committed:
Add docs for legacy impl

paulhauner committed:
Add more tests for legacy impl

paulhauner committed:
Panic tree hash derive if enum has multi variants

paulhauner committed:
Tidy committee cache

paulhauner committed:
Remove old macro

paulhauner committed:
Add test for decoding genesis state

paulhauner committed:
Fix shim in committee cache

paulhauner commented:
I've just added a fix to the `CommitteeCache` legacy serialization. I also rebased on `unstable` to fix a merge conflict.

michaelsproul commented on testing/ef_tests/src/cases/rewards.rs:
```diff
@@ -26,11 +27,16 @@ pub struct Deltas {
     source_deltas: Deltas,
     target_deltas: Deltas,
     head_deltas: Deltas,
+    #[ssz(with = "four_byte_option_deltas")]
```
I don't think this struct needs to implement `Decode` or `Encode`, I'm not sure why it previously did (everything seems to compile without it)

michaelsproul reviewed (approved):
This is impeccable! :ok_hand:

paulhauner committed:
Remove unchecked arith from ssz_derive

paulhauner committed:
Remove ssz impl from AllDeltas

paulhauner commented:
Thanks @michaelsproul! I've addressed your review comment in c9111f4.

I also made a small change in https://github.com/sigp/lighthouse/pull/2579/commits/91d3152f97ab0a2fff0927966531e370f46e97da to avoid unchecked arith that was triggering clippy in the `types` crate.

paulhauner commented:
Although I made a change since the last review, I believe the change is trivial. I'm keen to get this into `unstable` so we can tidy the `merge-f2f` branch and get it passing clippy. As such, I'm going to bors this. @michaelsproul could you please eyeball 91d3152, just in case? :pray: It's not going into prod until the merge, so it's not an issue in the meantime.

bors r+

paulhauner commented:
Looks like we're blocked on https://github.com/sigp/lighthouse/pull/2625 for the time being.

michaelsproul reviewed (approved):
I checked the arith and it looks good 👌 Merge at will. Could even batch with the zeroize PR?

paulhauner commented:
bors r+

paulhauner commented:
bors r+

njgheorghita commented:
@paulhauner :wave: It looks like the new versions of `eth2_ssz` / `eth2_ssz_derive` / `eth2_ssz_types` weren't published. Do you know if/when you expect them to be available?
